File: packages/react-native-fipa/ios/SessionVault.swift
import CryptoKit
import Foundation

/// No application extension/shared-access-group support. The native lock covers
/// every bridge/runtime in this app process, while the durable lease fences late
/// responses and survives a process restart.
enum VaultAccessibility: String, Codable {
  case whenUnlocked = "when-unlocked"
  case afterFirstUnlock = "after-first-unlock"
}

enum SessionVaultError: String, Error {
  case invalidInput = "vault_invalid_input"
  case busy = "vault_busy"
  case lostLease = "vault_lost_lease"
  case corrupt = "vault_corrupt"
  case locked = "vault_locked"
  case unavailable = "vault_unavailable"
  case storage = "vault_storage_failed"
  case policyMismatch = "vault_policy_mismatch"
}

protocol SessionVaultStorage {
  func read(account: String) throws -> Data?
  func write(account: String, data: Data, accessibility: VaultAccessibility) throws
}

struct VaultLease: Codable {
  let id: String
  var expiresAt: Double
  var preserveSession: Bool?
}

struct VaultRecord: Codable {
  let version: Int
  let account: String
  let accessibility: VaultAccessibility
  var generation: Int64
  var identityJSON: String?
  var sessionJSON: String?
  var recoverySessionJSON: String?
  var rollbackSessionJSON: String?
  var hasInteraction: Bool?
  var recoveryRequired: Bool
  var lease: VaultLease?
}

struct VaultSnapshot {
  let leaseId: String
  let generation: Int64
  let identityJSON: String?
  let sessionJSON: String?
  let recoveryRequired: Bool
}

final class SessionVault {
  // A static lock, not a per-module or per-JS-runtime lock. Keychain calls are
  // synchronous and bounded to local storage work. No network is performed here.
  private static let mutationLock = NSLock()
  static let maximumGeneration: Int64 = 9_007_199_254_740_991
  private let storage: SessionVaultStorage
  private let now: () -> Double
  private let random: () -> String

  init(storage: SessionVaultStorage,
       now: @escaping () -> Double = { Date().timeIntervalSince1970 },
       random: @escaping () -> String = {
         SymmetricKey(size: .bits256).withUnsafeBytes { Data($0) }
           .base64EncodedString().replacingOccurrences(of: "+", with: "-")
           .replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "")
       }) {
    self.storage = storage
    self.now = now
    self.random = random
  }

  func acquire(namespace: String, slot: String, accessibility: VaultAccessibility,
               leaseMilliseconds: Double, preserveSession: Bool = false) throws -> VaultSnapshot {
    let account = try Self.account(namespace: namespace, slot: slot)
    let duration = try Self.duration(leaseMilliseconds)
    return try locked {
      var record = try read(account) ?? VaultRecord(version: 1, account: account,
        accessibility: accessibility, generation: 0, identityJSON: nil,
        sessionJSON: nil, recoveryRequired: false, lease: nil)
      guard record.accessibility == accessibility else { throw SessionVaultError.policyMismatch }
      if let lease = record.lease {
        guard lease.expiresAt <= now() else { throw SessionVaultError.busy }
        // Only an interaction lease can recover independently committed access.
        // A refresh/session mutation never restores a potentially consumed token.
        try invalidate(&record, recover: lease.preserveSession == true)
      }
      let id = random()
      record.lease = VaultLease(id: id, expiresAt: now() + duration, preserveSession: preserveSession)
      try write(record)
      return VaultSnapshot(leaseId: id, generation: record.generation,
        identityJSON: record.identityJSON, sessionJSON: record.sessionJSON,
        recoveryRequired: record.recoveryRequired)
    }
  }

  /// A successful atomic commit consumes the lease. The caller must not expose
  /// newly issued access credentials before this write completes.
  func commit(namespace: String, slot: String, leaseId: String, generation: Double,
              identityJSON: String?, sessionJSON: String?, recoverySessionJSON: String? = nil,
              hasInteraction: Bool = false) throws -> Int64 {
    try Self.validateJSON(identityJSON, limit: 8192)
    try Self.validateJSON(sessionJSON, limit: 65536)
    try Self.validateJSON(recoverySessionJSON, limit: 65536)
    let account = try Self.account(namespace: namespace, slot: slot)
    let expected = try Self.generation(generation)
    return try locked {
      var record = try owned(account, leaseId: leaseId, generation: expected)
      record.generation = try successor(record.generation)
      record.rollbackSessionJSON = record.lease?.preserveSession == true ? record.recoverySessionJSON : nil
      record.identityJSON = identityJSON
      record.sessionJSON = sessionJSON
      record.recoverySessionJSON = recoverySessionJSON
      record.hasInteraction = hasInteraction
      record.recoveryRequired = false
      record.lease = nil
      try write(record)
      return record.generation
    }
  }

  func renew(namespace: String, slot: String, leaseId: String, generation: Double,
             leaseMilliseconds: Double) throws {
    let account = try Self.account(namespace: namespace, slot: slot)
    let expected = try Self.generation(generation)
    let duration = try Self.duration(leaseMilliseconds)
    try locked {
      var record = try owned(account, leaseId: leaseId, generation: expected)
      record.lease?.expiresAt = now() + duration
      try write(record)
    }
  }
  /// Release a read-only operation without discarding a known committed session.
  /// A preempted or expired lease still fails and cannot restore any old state.
  func release(namespace: String, slot: String, leaseId: String, generation: Double) throws {
    let account = try Self.account(namespace: namespace, slot: slot)
    let expected = try Self.generation(generation)
    try locked {
      var record = try owned(account, leaseId: leaseId, generation: expected)
      record.generation = try successor(record.generation)
      record.lease = nil
      try write(record)
    }
  }

  /// Persist irreversible key-registration progress without releasing the slot.
  /// Session material and generation stay unchanged while this lease owns it.
  func saveIdentity(namespace: String, slot: String, leaseId: String, generation: Double,
                    identityJSON: String) throws {
    try Self.validateJSON(identityJSON, limit: 8192)
    let account = try Self.account(namespace: namespace, slot: slot)
    let expected = try Self.generation(generation)
    try locked {
      var record = try owned(account, leaseId: leaseId, generation: expected)
      record.identityJSON = identityJSON
      try write(record)
    }
  }

  /// Clear capabilities before a best-effort remote termination, keeping this
  /// operation's lease and identity so another runtime cannot start a new login.
  func clearSession(namespace: String, slot: String, leaseId: String, generation: Double) throws {
    let account = try Self.account(namespace: namespace, slot: slot)
    let expected = try Self.generation(generation)
    try locked {
      var record = try owned(account, leaseId: leaseId, generation: expected)
      record.sessionJSON = nil
      record.recoverySessionJSON = nil
      record.rollbackSessionJSON = nil
      record.hasInteraction = false
      try write(record)
    }
  }

  /// Interaction failures may recover independent, previously committed access.
  /// Session mutations never restore a potentially consumed refresh token.
  func abandon(namespace: String, slot: String, leaseId: String, generation: Double) throws -> Int64 {
    let account = try Self.account(namespace: namespace, slot: slot)
    let expected = try Self.generation(generation)
    return try locked {
      var record = try owned(account, leaseId: leaseId, generation: expected)
      try invalidate(&record, recover: record.lease?.preserveSession == true)
      try write(record)
      return record.generation
    }
  }

  /// Logout/cancel can preempt an in-flight operation. Durable generation fences
  /// prevent any previously issued lease from writing its late result back.
  func invalidate(namespace: String, slot: String, accessibility: VaultAccessibility) throws -> Int64 {
    let account = try Self.account(namespace: namespace, slot: slot)
    return try locked {
      var record = try read(account) ?? VaultRecord(version: 1, account: account,
        accessibility: accessibility, generation: 0, identityJSON: nil,
        sessionJSON: nil, recoveryRequired: false, lease: nil)
      guard record.accessibility == accessibility else { throw SessionVaultError.policyMismatch }
      try invalidate(&record)
      try write(record)
      return record.generation
    }
  }

  /// Cancel only an interaction. A refresh lease must settle first; it may have
  /// consumed the recovery token. Return the old step only for best-effort remote
  /// cancellation, after durable local removal and generation fencing.
  func cancelInteraction(namespace: String, slot: String, accessibility: VaultAccessibility) throws -> (sessionJSON: String?, cancelled: Bool) {
    let account = try Self.account(namespace: namespace, slot: slot)
    return try locked {
      guard var record = try read(account) else { return (nil, false) }
      guard record.accessibility == accessibility else { throw SessionVaultError.policyMismatch }
      if let lease = record.lease, lease.expiresAt > now(), lease.preserveSession != true { throw SessionVaultError.busy }
      guard record.lease != nil || record.hasInteraction == true else { return (nil, false) }
      let previous = record.sessionJSON
      let recover = record.lease == nil || record.lease?.preserveSession == true
      try invalidate(&record, recover: recover)
      try write(record)
      return (previous, true)
    }
  }

  static func account(namespace: String, slot: String) throws -> String {
    guard !namespace.isEmpty, namespace.utf8.count <= 4096,
          !slot.isEmpty, slot.utf8.count <= 256 else { throw SessionVaultError.invalidInput }
    let bytes = try JSONEncoder().encode(["device-attestation-vault/v1", namespace, slot])
    return SHA256.hash(data: bytes).map { String(format: "%02x", $0) }.joined()
  }

  /// Compensate a cancelled commit only while its generation is still current.
  /// A later committed login must never be erased by this stale cleanup.
  func discard(namespace: String, slot: String, generation: Double) throws -> Bool {
    let account = try Self.account(namespace: namespace, slot: slot)
    let expected = try Self.generation(generation)
    return try locked {
      guard var record = try read(account), record.generation == expected else { return false }
      record.recoverySessionJSON = record.rollbackSessionJSON
      try invalidate(&record, recover: true)
      try write(record)
      return true
    }
  }

  private func owned(_ account: String, leaseId: String, generation: Int64) throws -> VaultRecord {
    guard let record = try read(account), record.generation == generation,
          let lease = record.lease, lease.id == leaseId, lease.expiresAt > now() else {
      throw SessionVaultError.lostLease
    }
    return record
  }
  private func read(_ account: String) throws -> VaultRecord? {
    guard let bytes = try storage.read(account: account) else { return nil }
    guard bytes.count <= 524_288,
          let record = try? JSONDecoder().decode(VaultRecord.self, from: bytes),
          record.version == 1, record.account == account,
          record.generation >= 0, record.generation <= Self.maximumGeneration,
          record.lease == nil || (record.lease!.id.utf8.count == 43 && record.lease!.expiresAt.isFinite) else {
      throw SessionVaultError.corrupt
    }
    do {
      try Self.validateJSON(record.identityJSON, limit: 8192)
      try Self.validateJSON(record.sessionJSON, limit: 65536)
      try Self.validateJSON(record.recoverySessionJSON, limit: 65536)
      try Self.validateJSON(record.rollbackSessionJSON, limit: 65536)
    } catch { throw SessionVaultError.corrupt }
    return record
  }
  private func write(_ record: VaultRecord) throws {
    let bytes = try JSONEncoder().encode(record)
    guard bytes.count <= 524_288 else { throw SessionVaultError.invalidInput }
    try storage.write(account: record.account, data: bytes, accessibility: record.accessibility)
  }
  private func invalidate(_ record: inout VaultRecord, recover: Bool = false) throws {
    record.generation = try successor(record.generation)
    record.sessionJSON = recover ? record.recoverySessionJSON : nil
    record.recoverySessionJSON = record.sessionJSON
    record.rollbackSessionJSON = nil
    record.hasInteraction = false
    record.recoveryRequired = true
    record.lease = nil
  }
  private func successor(_ generation: Int64) throws -> Int64 {
    guard generation < Self.maximumGeneration else { throw SessionVaultError.corrupt }
    return generation + 1
  }
  private static func generation(_ value: Double) throws -> Int64 {
    guard value.isFinite, value >= 0, value <= Double(maximumGeneration), value.rounded(.towardZero) == value else {
      throw SessionVaultError.invalidInput
    }
    return Int64(value)
  }
  private static func duration(_ milliseconds: Double) throws -> Double {
    guard milliseconds.isFinite, milliseconds >= 5000, milliseconds <= 120_000,
          milliseconds.rounded(.towardZero) == milliseconds else { throw SessionVaultError.invalidInput }
    return milliseconds / 1000
  }
  private static func validateJSON(_ value: String?, limit: Int) throws {
    guard let value else { return }
    guard value.utf8.count <= limit,
          let object = try? JSONSerialization.jsonObject(with: Data(value.utf8)),
          object is [String: Any] else { throw SessionVaultError.invalidInput }
  }
  private func locked<T>(_ body: () throws -> T) rethrows -> T {
    Self.mutationLock.lock()
    defer { Self.mutationLock.unlock() }
    return try body()
  }
}

File: packages/react-native-fipa/android/src/main/java/com/deviceattestation/SessionVault.kt
package com.deviceattestation

/**
 * Storage owns a lock spanning the entire read/modify/write transaction, across module instances
 * and processes. Network and signing are never performed here.
 */
internal interface VaultSlotStorage {
    val generationFloor: Long
        get() = 0

    val generationCeiling: Long
        get() = VaultCodec.MAX_GENERATION

    fun read(): ByteArray?

    fun write(bytes: ByteArray)
}

internal interface SessionVaultStorage {
    fun <T> transaction(account: String, operation: (VaultSlotStorage) -> T): T
}

internal data class VaultSnapshot(
    val leaseId: String,
    val generation: Long,
    val identityJSON: String?,
    val sessionJSON: String?,
    val recoveryRequired: Boolean,
)

internal data class CancelledInteraction(val sessionJSON: String?, val cancelled: Boolean)

internal class SessionVault(
    private val storage: SessionVaultStorage,
    private val now: () -> Long = System::currentTimeMillis,
    private val random: () -> String = FirstPartyCrypto::randomToken,
) {
    private fun successor(value: Long): Long {
        requireNative(value < VaultCodec.MAX_GENERATION, "vault_corrupt")
        return value + 1
    }

    private fun invalidate(record: VaultRecord, recover: Boolean = false) {
        record.generation = successor(record.generation)
        record.sessionJSON = if (recover) record.recoverySessionJSON else null
        record.recoverySessionJSON = record.sessionJSON
        record.rollbackSessionJSON = null
        record.hasInteraction = false
        record.recoveryRequired = true
        record.lease = null
    }

    private fun read(slot: VaultSlotStorage, account: String): VaultRecord? =
        slot.read()?.let {
            VaultCodec.decode(it, account).also { record ->
                requireNative(
                    record.generation in slot.generationFloor..slot.generationCeiling,
                    "vault_corrupt",
                )
            }
        }

    private fun write(slot: VaultSlotStorage, record: VaultRecord) {
        requireNative(
            record.generation in slot.generationFloor..slot.generationCeiling,
            "vault_corrupt",
        )
        slot.write(VaultCodec.encode(record))
    }

    private fun owned(
        slot: VaultSlotStorage,
        account: String,
        id: String,
        generation: Long,
    ): VaultRecord {
        val record = read(slot, account)
        requireNative(
            record != null &&
                record.generation == generation &&
                record.lease?.id == id &&
                record.lease!!.expiresAt > now(),
            "vault_lost_lease",
        )
        return record!!
    }

    fun acquire(
        namespace: String,
        slotId: String,
        accessibility: String,
        leaseMilliseconds: Double,
        preserveSession: Boolean,
    ): VaultSnapshot {
        val account = VaultCodec.account(namespace, slotId)
        VaultCodec.policy(accessibility)
        val duration = VaultCodec.duration(leaseMilliseconds)
        return storage.transaction(account) { slot ->
            val record =
                read(slot, account) ?: VaultRecord(account, generation = slot.generationFloor)
            record.lease?.let {
                requireNative(it.expiresAt <= now(), "vault_busy")
                invalidate(record, it.preserveSession)
            }
            val id = random()
            FirstPartyCrypto.decodeDigest(id)
            record.lease = VaultLease(id, now() + duration, preserveSession)
            write(slot, record)
            VaultSnapshot(
                id,
                record.generation,
                record.identityJSON,
                record.sessionJSON,
                record.recoveryRequired,
            )
        }
    }

    private fun <T> mutate(
        namespace: String,
        slotId: String,
        id: String,
        generation: Double,
        operation: (VaultRecord) -> T,
    ): T {
        val account = VaultCodec.account(namespace, slotId)
        val expected = VaultCodec.generation(generation)
        return storage.transaction(account) { slot ->
            val record = owned(slot, account, id, expected)
            val result = operation(record)
            write(slot, record)
            result
        }
    }

    fun commit(
        namespace: String,
        slotId: String,
        leaseId: String,
        generation: Double,
        identityJSON: String?,
        sessionJSON: String?,
        recoverySessionJSON: String?,
        hasInteraction: Boolean,
    ): Long {
        VaultCodec.validateJSON(identityJSON, 8192)
        VaultCodec.validateJSON(sessionJSON, 65536)
        VaultCodec.validateJSON(recoverySessionJSON, 65536)
        return mutate(namespace, slotId, leaseId, generation) { record ->
            record.generation = successor(record.generation)
            record.rollbackSessionJSON =
                if (record.lease!!.preserveSession) record.recoverySessionJSON else null
            record.identityJSON = identityJSON
            record.sessionJSON = sessionJSON
            record.recoverySessionJSON = recoverySessionJSON
            record.hasInteraction = hasInteraction
            record.recoveryRequired = false
            record.lease = null
            record.generation
        }
    }

    fun renew(
        namespace: String,
        slot: String,
        id: String,
        generation: Double,
        leaseMilliseconds: Double,
    ) {
        val duration = VaultCodec.duration(leaseMilliseconds)
        mutate(namespace, slot, id, generation) { it.lease!!.expiresAt = now() + duration }
    }

    fun release(namespace: String, slot: String, id: String, generation: Double) {
        mutate(namespace, slot, id, generation) {
            it.generation = successor(it.generation)
            it.lease = null
        }
    }

    fun saveIdentity(
        namespace: String,
        slot: String,
        id: String,
        generation: Double,
        identity: String,
    ) {
        VaultCodec.validateJSON(identity, 8192)
        mutate(namespace, slot, id, generation) { it.identityJSON = identity }
    }

    fun clearSession(namespace: String, slot: String, id: String, generation: Double) {
        mutate(namespace, slot, id, generation) {
            it.sessionJSON = null
            it.recoverySessionJSON = null
            it.rollbackSessionJSON = null
            it.hasInteraction = false
        }
    }

    fun abandon(namespace: String, slot: String, id: String, generation: Double): Long =
        mutate(namespace, slot, id, generation) {
            invalidate(it, it.lease!!.preserveSession)
            it.generation
        }

    fun invalidate(namespace: String, slotId: String, accessibility: String): Long {
        val account = VaultCodec.account(namespace, slotId)
        VaultCodec.policy(accessibility)
        return storage.transaction(account) { slot ->
            val record =
                read(slot, account) ?: VaultRecord(account, generation = slot.generationFloor)
            invalidate(record)
            write(slot, record)
            record.generation
        }
    }

    fun cancelInteraction(
        namespace: String,
        slotId: String,
        accessibility: String,
    ): CancelledInteraction {
        val account = VaultCodec.account(namespace, slotId)
        VaultCodec.policy(accessibility)
        return storage.transaction(account) { slot ->
            val record = read(slot, account) ?: return@transaction CancelledInteraction(null, false)
            val lease = record.lease
            requireNative(
                lease == null || lease.expiresAt <= now() || lease.preserveSession,
                "vault_busy",
            )
            if (lease == null && !record.hasInteraction)
                return@transaction CancelledInteraction(null, false)
            val previous = record.sessionJSON
            invalidate(record, lease == null || lease.preserveSession)
            write(slot, record)
            CancelledInteraction(previous, true)
        }
    }

    fun discard(namespace: String, slotId: String, generation: Double): Boolean {
        val account = VaultCodec.account(namespace, slotId)
        val expected = VaultCodec.generation(generation)
        return storage.transaction(account) { slot ->
            val record = read(slot, account) ?: return@transaction false
            if (record.generation != expected) return@transaction false
            record.recoverySessionJSON = record.rollbackSessionJSON
            invalidate(record, true)
            write(slot, record)
            true
        }
    }
}
